### ISSUE TITLE
support Fulltext Indexing by Entity's status

### ISSUE THREAD

**Issue by lmtoo** [paulcwarren/spring-content#138]
Hi @paulcwarren , I have a problem , fulltext by entity's status, this entity already make association with content, but by this framework fulltext right after the content set. Is a avaliable provide a trigger to entity's content Indexing ?

**Comment by paulcwarren** [comment]
So, rather than automatically fulltext index content immediately after it is set, you want to delay fulltext indexing until some later time of your choosing?  i.e. when an entity's status changes from "under review" to "approved" for example?  

Is that right?  

I'm guessing your entity is being routed around some sort of workflow?

**Comment by lmtoo** [comment]
yes, that right

**Comment by paulcwarren** [comment]
Hi @lmtoo I was thinking about this request and I have a proposal that I wanted to run past you. 

Atm the fulltext indexing capability isn't a (spring) service yet.  The indexing logic for both Elasticsearch and SolrSearch (if I recall you are using elastic) are implemented directly in `afterSetContent` store event handlers.  Not very useful for other consumers.  I would be more than happy.  In fact, I think it is a better design anyway, to refactor that logic into a Elastic and Solr implementation of a `FulltextIndexingService` and have the existing store event handlers use that service instead of implement it themselves.  This would provide you with a service you could autowire into your own code.

On the other hand I don't think there is much point in Spring Content having some sort of customizable fulltext triggering f/w; i.e. the logic that detects change in entity "status" and calls the fulltext indexing service.  There are plenty of good trigger mechanisms in spring (aop, data rest event handlers, ...) and jpa (listeners) so I think the best alternative here is for you to implement that piece yourself using one of these existing frameworks.  I would recommend using spring aop personally but you could use any.  Spring aop might look something like this:-

```
@Aspect
public class IndexingAspect {
 
   @Autowired
    private MyEntityRepository repo;

   @Autowired
    private MyEntityStore store;

   @Autowired
   private FulltextIndexingService ftiService;

   @Around("execution(* my.CrudRepository.save(..))") 
   public void index(JoinPoint point) {

       MyEntity entityBeingSaved = (MyEntity)point.getArgs()[0];
       Status entityBeingSavedStatus = entityBeingSaved.getStatus();
       MyEntity existingEntity = repo.findById(entityBeingSaved.getId())
       Status existingStatus = existingEntity.getStatus();
       
       MyEntity savedEntity = point.proceed();

       if (entityBeingSavedStatus != existingStatus) {
          ftiService.index(store.getContent(savedEntity));
       
   }
}
```    

Does this sound reasonable to you?

**Comment by lmtoo** [comment]
FulltextIndexingService is well, I can't wait to use this new feature

**Comment by paulcwarren** [comment]
Fixed in this [commit](https://github.com/paulcwarren/spring-content/commit/2cf2873115d3709923c95e663acf53428ccafff1).

I added brief documentation [here](https://paulcwarren.github.io/spring-content/refs/snapshot/1.0.x/elasticsearch-index.html#_custom_indexing).  It should work as shown above.

I will cut 1.0.0.M8 for you tomorrow.

**Comment by lmtoo** [comment]
Hi @paulcwarren , I have seen this new feature just now ,but I can't close fulltextIndex right after the content set

**Comment by paulcwarren** [comment]
Hi @lmtoo sorry to hear that.  

I added [this test](https://github.com/paulcwarren/spring-content/blob/dc6af7f9b9804991c6b320485e002a78731ed5a6/spring-content-elasticsearch/src/test/java/org/springframework/content/elasticsearch/ElasticsearchIT.java#L152).

It was green in [this build](https://travis-ci.org/github/paulcwarren/spring-content/builds/671947436) so I wanted to check my understanding of the issue.  How does your case differ from the test?

**Comment by lmtoo** [comment]
Hi, @paulcwarren  I'm sorry. I mean disable automatically fulltext index by config application.properties. so I can index by invoking IndexService method.

**Comment by paulcwarren** [comment]
I see.  I had wondered if you wanted to do that.  I should have asked.  Make sense.  I will see what we can do.  Should be easy I think.

**Comment by paulcwarren** [comment]
@lmtoo do you use spring boot?  spring boot offers much better support for properties and conditional beans than regular spring.  We don't have a spring boot starter for elasticsearch but we should have one so I am thinking of implementing the new enable/disable property in a new `spring-content-elasticsearch-boot-starter`.  would that work for you?  or do you not use spring boot?

**Comment by paulcwarren** [comment]
Hi @lmtoo.  This [commit](https://github.com/paulcwarren/spring-content/commit/5a067e592494f2fa3dd5e122c53cbb4078109d4c) added a spring-content-elasticsearch-boot-starter that includes a `spring.content.elasticsearch.autoindex` property allowing you to disable auto indexing and is available in the latest 1.0.0.M10-SNAPSHOT.
